Lib/test/test_clinic.py

```diff
@@ -2863,6 +2877,121 @@ def test_cloned_func_with_converter_exception_message(self):
                 func = getattr(ac_tester, name)
                 self.assertEqual(func(), name)
 
+    def test_depr_star_pos0_len1(self):
+        fn = ac_tester.depr_star_pos0_len1
+        self.check_depr_star(
+            fname="depr_star_pos0_len1", pnames="'a'",
+            good_calls=[
+                lambda: fn(a=None),
+            ],
+            bad_calls=[
+                lambda: fn(None),
+            ])
+
+    def test_depr_star_pos0_len2(self):
+        fn = ac_tester.depr_star_pos0_len2
+        self.check_depr_star(
+            fname="depr_star_pos0_len2", pnames="'a' and 'b'",
+            good_calls=[
+                lambda: fn(a=None, b=None),
+            ],
+            bad_calls=[
+                lambda: fn(None, b=None),
+                lambda: fn(None, None),
+            ])
+
+    def test_depr_star_pos0_len3_with_kwd(self):
+        fn = ac_tester.depr_star_pos0_len3_with_kwd
+        self.check_depr_star(
+            fname="depr_star_pos0_len3_with_kwd", pnames="'a', 'b' and 'c'",
+            good_calls=[
+                lambda: fn(a=None, b=None, c=None, d=None),
+            ],
+            bad_calls=[
+                lambda: fn(None, b=None, c=None, d=None),
+                lambda: fn(None, None, c=None, d=None),
+                lambda: fn(None, None, None, d=None),
+            ])
```

Comment:
Why not simple
```python
fn(a=None, b=None, c=None, d=None)
self.check_depr_star(lambda: fn(None, b=None, c=None, d=None))
self.check_depr_star(lambda: fn(None, None, c=None, d=None))
self.check_depr_star(lambda: fn(None, None, None, d=None))
```
or even
```python
fn(a=None, b=None, c=None, d=None)
self.check_depr_star(fn, None, b=None, c=None, d=None)
self.check_depr_star(fn, None, None, c=None, d=None)
self.check_depr_star(fn, None, None, None, d=None)
```
?

Every time I see a test helper which takes a list of cases, I think that it would be clearer with separate calls for every case. At least you will have a traceback pointing to the failed case.

Reply:
Yes, the traceback will be better with this approach. I'll adjust the PR.

Reply:
We will have to simplify the regex, though. Perhaps that is an acceptable tradeoff. We could match with `fn.__name__` for the function name, but it would be nice to have a way to match against the parameter names also. Perhaps something like:

```python
self.check_depr_star(fn, "a", b=0, c=0, d=0)
```

... and then use `*args` to check for the param names in the deprecation message.

Reply:
See 6559443c7b028fb309d800006f42cb3ce1f123ac. I really like how this turned out; your suggested approach is a lot more readable, @serhiy-storchaka. What do you think, @AlexWaygood?

Reply:
Yes, but I would switch `fn` and `pnames` for more natural look.
```python
self.check_depr_star("a", fn, b=0, c=0, d=0)
```

Reply:
I'm not sure that will work out as nice as 6559443c7b028fb309d800006f42cb3ce1f123ac. In that commit I manipulate `*args` to produce the `pnames` regex part (I suspect @AlexWaygood is not very fond of that part). I think we should keep `*args` for readability, since it harmonises well with how the call would look. We could pass in pnames explicitly before `fn`, but we would have to do that for every `check_depr_star` call. I believe 6559443c7b028fb309d800006f42cb3ce1f123ac is a good compromise.

Reply:
> ```python
> self.check_depr_star("'a', 'b' and 'c'", fn, 1, b=2, c=3, d=4)
> ```

Yes, that's the option I hinted to in https://github.com/python/cpython/pull/107768#discussion_r1288365103. Still, we would have to supply that for every call. Perhaps that is better.

Reply:
Thinking more about it, we'll have to use the explicit pnames variant; we'll end up constructing the wrong regex in some cases. I'll amend the PR.

Reply:
See 194420bcd15db20e961352aa43fbc988dd1457b3

Reply:
How do you get "and" between parameter names without reimplementing `pprint_words()` from `clinic.py`?
```python
check = partial(self.check_depr_star, ("'a', 'b' and 'c'")
check(fn, ...)
```
I have not noticed it before, but it seems that `* [from ...]` is inserted before the first parameter. You approach does not work if there are still some positional parameter left.

Reply:
> How do you get "and" between parameter names without reimplementing `pprint_words()` from `clinic.py`?
> 
> ```python
> check = partial(self.check_depr_star, "'a', 'b' and 'c'")
> check(fn, ...)
> ```
> 
> I have not noticed it before, but it seems that `* [from ...]` is inserted before the first parameter. You approach does not work if there are still some positional parameter left.

See https://github.com/python/cpython/pull/107768#discussion_r1288374764 for answers.

Reply:
Thanks, Serhiy. I think this turned out very well; the new helper API makes the tests a lot more readable. Very neat.